apollo-router/src/configuration/cors.rs

```diff
@@ -84,91 +271,371 @@ impl Cors {
         allow_credentials: Option<bool>,
         allow_headers: Option<Vec<String>>,
         expose_headers: Option<Vec<String>>,
-        origins: Option<Vec<String>>,
-        match_origins: Option<Vec<String>>,
-        methods: Option<Vec<String>>,
         max_age: Option<Duration>,
+        methods: Option<Vec<String>>,
+        policies: Option<Vec<Policy>>,
     ) -> Self {
         Self {
-            expose_headers,
-            match_origins,
-            max_age,
-            origins: origins.unwrap_or_else(default_origins),
-            methods: methods.unwrap_or_else(default_cors_methods),
             allow_any_origin: allow_any_origin.unwrap_or_default(),
             allow_credentials: allow_credentials.unwrap_or_default(),
             allow_headers: allow_headers.unwrap_or_default(),
+            expose_headers,
+            max_age,
+            methods: methods.unwrap_or_else(default_cors_methods),
+            policies: policies.or_else(|| Some(vec![Policy::default()])),
         }
     }
 }
 
-impl Cors {
-    pub(crate) fn into_layer(self) -> Result<CorsLayer, String> {
+/// Our custom CORS layer that supports per-origin configuration
+#[derive(Clone, Debug)]
+pub(crate) struct CorsLayer {
```

Comment:
We should move the `CorsLayer` / `CorsService` implementations out of the `configuration` module